pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1426,3 +1444,258 @@ func (vs *VSphere) GetZoneByNodeName(ctx context.Context, nodeName k8stypes.Node
 func (vs *VSphere) GetZoneByProviderID(ctx context.Context, providerID string) (cloudprovider.Zone, error) {
 	return cloudprovider.Zone{}, cloudprovider.NotImplemented
 }
+
+// GetLabelsForVolume implements the PVLabeler interface for VSphere
+// since this interface is used by the PV label admission controller.
+func (vs *VSphere) GetLabelsForVolume(ctx context.Context, pv *v1.PersistentVolume) (map[string]string, error) {
+	// ignore if not vSphere volume
+	if pv.Spec.VsphereVolume == nil {
+		return nil, nil
+	}
+	return vs.GetVolumeLabels(pv.Spec.VsphereVolume.VolumePath)
+}
+
+// GetVolumeLabels returns the well known zone and region labels for given volume
+func (vs *VSphere) GetVolumeLabels(volumePath string) (map[string]string, error) {
+	// Create context
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	datastorePathObj, err := vclib.GetDatastorePathObjFromVMDiskPath(volumePath)
+	if err != nil {
+		klog.Errorf("Failed to get datastore for volume: %v: %+v", volumePath, err)
+		return nil, err
+	}
+	dsZones, err := vs.GetZonesForDatastore(ctx, datastorePathObj.Datastore)
+	if err != nil {
+		klog.Errorf("Failed to get zones for datastore %v: %+v", datastorePathObj.Datastore, err)
+		return nil, err
+	}
+	dsZones, err = vs.collapseZonesInRegion(ctx, dsZones)
+	// FIXME: For now, pick the first zone of datastore as the zone of volume
+	labels := make(map[string]string)
+	if len(dsZones) > 0 {
+		labels[kubeletapis.LabelZoneRegion] = dsZones[0].Region
+		labels[kubeletapis.LabelZoneFailureDomain] = dsZones[0].FailureDomain
+	}
+	return labels, nil
+}
+
+// collapse all zones in same region. Join FailureDomain with well known separator
+func (vs *VSphere) collapseZonesInRegion(ctx context.Context, zones []cloudprovider.Zone) ([]cloudprovider.Zone, error) {
+	// first create a map of region -> list of zones in that region
+	regionToZones := make(map[string][]string)
+	for _, zone := range zones {
+		fds, exists := regionToZones[zone.Region]
+		if !exists {
+			fds = make([]string, 0)
+		}
+		regionToZones[zone.Region] = append(fds, zone.FailureDomain)
+	}
+
+	// Join all fds in same region and return Zone instances
+	collapsedZones := make([]cloudprovider.Zone, 0)
+	for region, fds := range regionToZones {
+		fdSet := sets.NewString(fds...)
+		appendedZone := volumeutil.ZonesSetToLabelValue(fdSet)
```

Comment:
You may have to rebase this PR, a bunch of the volume zone helper methods were moved from `pkg/volume/util` to `k8s.io/cloud-provider/volume/helpers`.

Reply:
Thanks for the notice. Rebased.